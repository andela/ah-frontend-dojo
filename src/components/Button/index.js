// React libraries
import React from 'react';

// third-party libraries
import PropTypes from 'prop-types';

const Button = (props) => {
  // Destructure props
  const {
    btnClass, btnName, btnEvent, disabled,
  } = props;

  return (

    <button
      className={btnClass}
      type="button"
      onClick={btnEvent}
      disabled={disabled}
    >
      {btnName}
    </button>
  );
};

Button.defaultProps = {
  disabled: false,
};

// Props validation
Button.propTypes = {
  btnName: PropTypes.string.isRequired,
  btnClass: PropTypes.string.isRequired,
  btnEvent: PropTypes.func.isRequired,
  disabled: PropTypes.bool,

};
export default Button;
