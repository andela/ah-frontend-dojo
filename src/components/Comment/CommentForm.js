import React, { Component } from 'react';

// third party libraries
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

// comment actions
import createComment from 'store/actions/commentActions';
import Button from 'components/Button';

export class CommentForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      commentBody: '',
      error: '',
    };
  }

  handleCommentSubmit = (e) => {
    e.preventDefault();


    const { commentBody } = this.state;
    const { createNewComment, articleSlug } = this.props;

    if (articleSlug && commentBody) {
      createNewComment(articleSlug, commentBody);
      this.setState({ commentBody: '' });
    }
  };

  handleChange = (e) => {
    const { name, value } = e.target;

    this.setState({ [name]: value });
  };

  renderCommentError = error => (
    error ? <div className="error-msg text-danger">{ error }</div> : ''
  );

  renderCommentSuccess = success => (success ? <div className="success-msg text-success">Comment added successfully</div> : '');


  render() {
    const { commentBody, error } = this.state;
    const { success, username } = this.props;

    return (
      <div className="comment_form">
        {this.renderCommentError(error)}
        {this.renderCommentSuccess(success)}

        <form onSubmit={this.handleCommentSubmit}>
          <div>
            <textarea
              placeholder="Comment"
              rows="3"
              required
              name="commentBody"
              value={commentBody}
              onChange={this.handleChange}
            />
          </div>
          <div>
            <Button inputType="submit" btnClass="btn" btnEvent={this.handleCommentSubmit} btnName="Post Comment" disabled={!username} />
            {!username && (
              <i>
                {' login to comment'}
              </i>
            )}
          </div>
        </form>
      </div>
    );
  }
}

// default props
CommentForm.defaultProps = {
  username: '',
  success: false,
};

// Props Validation
CommentForm.propTypes = {
  createNewComment: PropTypes.func.isRequired,
  articleSlug: PropTypes.string.isRequired,
  success: PropTypes.bool,
  username: PropTypes.string,
};

// mapState to props
export const mapStateToProps = (state) => {
  const { commentReducer: { success }, loginReducer: { user: { username } } } = state;
  return { success, username };
};

// map dispatch to props
export const mapDispatchToProps = dispatch => ({
  createNewComment(articleSlug, commentBody) {
    dispatch(createComment(articleSlug, commentBody));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(CommentForm);
