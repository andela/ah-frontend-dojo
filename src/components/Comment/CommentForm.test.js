import React from 'react';
import { shallow } from 'enzyme';
import Button from 'components/Button';
import { CommentForm, mapStateToProps, mapDispatchToProps } from './CommentForm';


describe('Comment component', () => {
  const props = {
    articleSlug: '',
    createNewComment: jest.fn(),
    username: 'dojo',
  };

  let wrapper;


  beforeEach(() => {
    wrapper = shallow(<CommentForm {...props} />);
  });


  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should limit commenting on articles to authenticated users', () => {
    expect(wrapper.find(Button).props().disabled).toBe(false);
    wrapper.setProps({ username: '' });
    expect(wrapper.find('i').text()).toBe(' login to comment');
    expect(wrapper.find(Button).props().disabled).toBe(true);
  });

  it('should handle the comment body input change event', () => {
    const event = {
      target: {
        name: 'commentBody',
        value: 'My first comment',
      },
    };

    wrapper.instance().handleChange(event);
    expect(wrapper.instance().state.commentBody).toBe(event.target.value);
  });

  it('should handle the onSubmit event', () => {
    const instance = wrapper.instance();
    wrapper.setState({ commentBody: '' });
    wrapper.setProps({ articleSlug: 'my_article_slug' });
    const event = {
      target: {
        type: 'submit',
        name: 'addComment',
      },
      preventDefault: jest.fn(),
    };

    instance.handleCommentSubmit(event);

    expect(instance.props.createNewComment.mock.calls.length).toBe(0);

    wrapper.setState({ commentBody: 'My first comment' });
    instance.handleCommentSubmit(event);

    expect(instance.props.createNewComment.mock.calls.length).toBe(1);
    expect(instance.props.createNewComment).toBeCalled();
  });

  it('should show an error on submitting a comment with  on missing comment body', () => {
    wrapper.setState({ error: 'Comment body is required' });
    expect(wrapper.find('.error-msg').text()).toBe('Comment body is required');
  });

  it('should show a success message on successfully commenting on an article', () => {
    wrapper.setProps({ success: true });
    expect(wrapper.find('.success-msg').text()).toBe('Comment added successfully');
  });


  it('should map state to props', () => {
    const mockedState = {
      loginReducer: {
        user: {
          username: 'arthur'
        }
      },
      commentReducer: {
        success: true,
      }
    };

    const state = mapStateToProps(mockedState);

    expect(state.username).toEqual('arthur');
    expect(state.success).toEqual(true);
  });

  it('should map dispatch to props', () => {
    const mockedDispatch = jest.fn();

    mapDispatchToProps(mockedDispatch).createNewComment();
    expect(mockedDispatch).toHaveBeenCalled();
  });
});
