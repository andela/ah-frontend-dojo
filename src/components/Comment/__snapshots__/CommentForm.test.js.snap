// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component should render without crashing 1`] = `
<div
  className="comment_form"
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <textarea
        name="commentBody"
        onChange={[Function]}
        placeholder="Comment"
        required={true}
        rows="3"
        value=""
      />
    </div>
    <div>
      <Button
        btnClass="btn"
        btnEvent={[Function]}
        btnName="Post Comment"
        disabled={false}
        inputType="submit"
      />
    </div>
  </form>
</div>
`;
